(GH-616) Capture 7z's stdout in Get-ChocolateyUnzip

Erlkoenig90 opened the pull request:
Sometimes, Install-ChocolateyZipPackage and Get-ChocolateyUnzip capture
files that don't belong to the package, record them in the
<packagename>Install.zip.txt file, and cause them to be removed upon
uninstallation by UnInstall-ChocolateyZipPackage. This fix uses 7z's
standard output instead to determine which files belong to the package.
This also avoids scanning possibly large directory trees.
This commit should fix the issue GH-616.

ferventcoder commented on src/chocolatey.resources/helpers/functions/Get-ChocolateyUnzip.ps1:
```diff
@@ -91,32 +91,44 @@ param(
+        Add-Content $zipExtractLogFullPath $file
+      }
+      # Print the line, such that it looks as if stdout was not redirected
+      Write-Debug $line
```
Would you move this to Write-Verbose? This will help with #476

ferventcoder commented on src/chocolatey.resources/helpers/functions/Get-ChocolateyUnzip.ps1:
```diff
@@ -91,32 +91,44 @@ param(
+        # This is a line indicating an extracted file
+        $file = $destination + "\" + $line.Substring(12)
+        # Save the filename
+        Add-Content $zipExtractLogFullPath $file
```
It might be better to hold the output in a variable and write the file once instead at the end.

ferventcoder replied:
There may be cases where the file doesn't exist as well. So it may be best not to write anything if the variable is not empty.

ferventcoder commented on src/chocolatey.resources/helpers/functions/Get-ChocolateyUnzip.ps1:
```diff
@@ -91,32 +91,44 @@ param(
 
-  if ($zipExtractLogFullPath) {
-    Write-Debug "wrapping 7za invocation with Write-FileUpdateLog"
-    Write-FileUpdateLog -logFilePath $zipExtractLogFullPath -locationToMonitor $destination -scriptToRun $unzipOps -argumentList $7zip,$destination32,$fileFullPath32,([ref]$exitCode)
```
This may fix #156 as well since we won't be using it anymore. :+1:

ferventcoder commented:
Overall :+1:

Erlkoenig90 commented:
Thanks for the quick feedback! I incorporated your suggestions in commit 4a2b56d086655e5a2181f0f473cccb1c97e0846e. The file is written using a StringBuffer and a FileStream, and removed again if nothing was written at all.

ferventcoder commented:
FileStream may have issues.

Erlkoenig90 commented:
Okay, what kind of issues? Should is change it to Add-Content?

ferventcoder commented:
It may be fine. No worries

ferventcoder commented on src/chocolatey.resources/helpers/functions/Get-ChocolateyUnzip.ps1:
```diff
@@ -91,32 +91,73 @@ param(
+    while(($process.StandardOutput -ne $null) -and (($line = $process.StandardOutput.ReadLine()) -ne $null)) {
+      if($line.StartsWith("Extracting")) {
+        # This is a line indicating an extracted file
+        $file = $destination + "\" + $line.Substring(12)
```
What is this `12`?

ferventcoder replied:
Nevermind, the two spaces after extracting.

Erlkoenig90 replied:
In 7z's output in the form "Extracting  path\of\file\in\zip", the file path begins at character 12.

ferventcoder commented:
I changed this a bit to better match Start-ProcessAsAdmin (it was pretty close already). I think you'll find the perf quite sufficient.

ferventcoder commented:
Commits squashed, rebased and updated at https://github.com/chocolatey/choco/commit/bde0941d7124d1aee4f5fc4ae5cc623ac4d3e389

Merged into stable at https://github.com/chocolatey/choco/commit/95c000d21f3594b203f7732f6e81bc451f65944a

ferventcoder commented:
Thanks for your contribution!
